pkg/syncclient/sync_client.go

```diff
@@ -21,101 +21,179 @@ import (
 
 	log "github.com/Sirupsen/logrus"
 
+	"context"
+	"sync"
+
+	"fmt"
+
 	"github.com/projectcalico/libcalico-go/lib/backend/api"
 	"github.com/projectcalico/typha/pkg/syncproto"
 )
 
+var nextID uint64
+
 func New(addr string, myVersion, myHostname, myInfo string, cbs api.SyncerCallbacks) *SyncerClient {
+	id := nextID
+	nextID++
 	return &SyncerClient{
+		ID:        id,
+		logCxt:    log.WithField("connID", id),
 		callbacks: cbs,
 		addr:      addr,
 
 		myVersion:  myVersion,
 		myHostname: myHostname,
 		myInfo:     myInfo,
+
+		PanicOnFailure: true,
 	}
 }
 
 type SyncerClient struct {
+	ID                            uint64
+	logCxt                        *log.Entry
 	callbacks                     api.SyncerCallbacks
 	addr                          string
 	myHostname, myVersion, myInfo string
+	c                             net.Conn
+	Finished                      sync.WaitGroup
+
+	PanicOnFailure bool
 }
 
 func (s *SyncerClient) Start() {
-	go s.loop()
+	err := s.StartContext(context.Background())
+	if err != nil {
+		s.logCxt.WithError(err).Panic("Failed to connect to Typha")
+	}
+}
+
+func (s *SyncerClient) StartContext(cxt context.Context) error {
```

Comment:
I think this is the convention for the "Contextified" version of an existing function but now I can't find an example...